src/bloqade/pyqrack/squin/runtime.py

```diff
@@ -0,0 +1,323 @@
+        else:
+            return [self.factor, 0, 0, self.factor]
+
+    def apply(self, *qubits: PyQrackQubit, adjoint: bool = False) -> None:
+        self.op.apply(*qubits, adjoint=adjoint)
+
+        target = qubits[-1]
+
+        # NOTE: just factor * eye(2)
+        m = self.mat(adjoint)
+
+        # TODO: output seems to always be normalized -- no-op?
+        target.sim_reg.mtrx(m, target.addr)
+
+    def control_apply(self, *qubits: PyQrackQubit, adjoint: bool = False) -> None:
+        self.op.control_apply(*qubits, adjoint=adjoint)
+
+        target = qubits[-1]
+        ctrls = [qbit.addr for qbit in qubits[:-1]]
+
+        m = self.mat(adjoint=adjoint)
+
+        target.sim_reg.mcmtrx(ctrls, m, target.addr)
+
+
+@dataclass(frozen=True)
+class MtrxOpRuntime(OperatorRuntimeABC):
+    def mat(self, adjoint: bool) -> list[complex]:
+        raise NotImplementedError("Override this method in the subclass!")
+
+    def apply(self, *qubits: PyQrackQubit, adjoint: bool = False) -> None:
+        target = qubits[-1]
+        m = self.mat(adjoint=adjoint)
+        target.sim_reg.mtrx(m, target.addr)
+
+    def control_apply(self, *qubits: PyQrackQubit, adjoint: bool = False) -> None:
+        target = qubits[-1]
+        ctrls = [qbit.addr for qbit in qubits[:-1]]
+
+        m = self.mat(adjoint=adjoint)
+
+        target.sim_reg.mcmtrx(ctrls, m, target.addr)
+
+
+@dataclass(frozen=True)
+class SpRuntime(MtrxOpRuntime):
+    def mat(self, adjoint: bool) -> list[complex]:
+        if adjoint:
+            return [0, 0, 0.5, 0]
+        else:
+            return [0, 0.5, 0, 0]
+
+
+@dataclass(frozen=True)
+class SnRuntime(MtrxOpRuntime):
+    def mat(self, adjoint: bool) -> list[complex]:
+        if adjoint:
+            return [0, 0.5, 0, 0]
+        else:
+            return [0, 0, 0.5, 0]
+
+
+@dataclass(frozen=True)
+class PhaseOpRuntime(MtrxOpRuntime):
+    theta: float
+    global_: bool
+
+    def mat(self, adjoint: bool) -> list[complex]:
+        sign = (-1) ** (not adjoint)
+        local_phase = np.exp(sign * 1j * self.theta)
+
+        # NOTE: this is just 1 if we want a local shift
+        global_phase = np.exp(sign * 1j * self.theta * self.global_)
+
+        return [global_phase, 0, 0, local_phase]
+
+
+@dataclass(frozen=True)
+class RotRuntime(OperatorRuntimeABC):
+    axis: OperatorRuntimeABC
+    angle: float
+
+    def apply(self, *qubits: PyQrackQubit, adjoint: bool = False) -> None:
+        sign = (-1) ** adjoint
+        angle = sign * self.angle
+        target = qubits[-1]
+
+        if not isinstance(self.axis, OperatorRuntime):
+            raise RuntimeError(
+                f"Rotation only supported for Pauli operators! Got {self.axis}"
+            )
+
+        try:
+            axis = getattr(Pauli, "Pauli" + self.axis.method_name.upper())
+        except KeyError:
+            raise RuntimeError(
+                f"Rotation only supported for Pauli operators! Got {self.axis}"
+            )
```

Comment:
Since this is a runtime specific to PyQrack you can put this check as a validation inside the `__post_init__` of this class and save the result as `pyqrack_axis` field